sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -547,6 +551,39 @@ public CompletionStage<MutateRowResponse> writeRecord(KV<ByteString, Iterable<Mu
 
       RowMutationEntry entry = RowMutationEntry.createFromMutationUnsafe(record.getKey(), mutation);
 
+      ServiceCallMetric serviceCallMetric = createServiceCallMetric();
+
+      CompletableFuture<MutateRowResponse> result = new CompletableFuture<>();
+
+      outstandingMutations += 1;
+      Futures.addCallback(
+          new VendoredListenableFutureAdapter<>(bulkMutation.add(entry)),
+          new WriteMutationCallback(result, serviceCallMetric),
+          directExecutor());
+      return result;
+    }
+
+    @Override
+    public CompletionStage<MutateRowResponse> writeSingleRecord(
+        KV<ByteString, Iterable<Mutation>> record) {
+      com.google.cloud.bigtable.data.v2.models.Mutation mutation =
+          com.google.cloud.bigtable.data.v2.models.Mutation.fromProtoUnsafe(record.getValue());
+
+      RowMutation rowMutation = RowMutation.create(tableId, record.getKey(), mutation);
+
+      ServiceCallMetric serviceCallMetric = createServiceCallMetric();
+
+      CompletableFuture<MutateRowResponse> result = new CompletableFuture<>();
+
+      outstandingMutations += 1;
+      Futures.addCallback(
+          new VendoredListenableFutureAdapter<>(client.mutateRowAsync(rowMutation)),
+          new WriteMutationCallback(result, serviceCallMetric),
+          directExecutor());
+      return result;
```

Comment:
why bother converting this to a ListenableFuture? its only used internally. I think it would be better to just stick with ApiFutures.

(separately I really wish someone would make ApiFuture implement CompleteableFuture :()

Reply:
Thats what we are using currently, and I followed that pattern. It looks like futures.addCallback takes a listenable future